_posts/2021-11-29-plugins-intro.markdown

```diff
@@ -0,0 +1,149 @@
+---
+layout: post
+title: "Introduction to OpenSearch Plugins"
+authors: 
+  - vemsarat
+  - kazabdu
+date: 2021-11-22
+categories:
+  - technical-post
+twittercard:
+  description: "This post walks through how plugins work in OpenSearch and how a custom Java Security Manager handle the security of OpenSearch."
+---
+
+
+OpenSearch enables extending core features via plugins. Plugins are empowered to access all extensible features of OpenSearch and extend them. In this blog post we wanted to unbox the plugin architecture and help understand how they work.
+
+
+## Pluggable Architecture
+
+Plugins in OpenSearch bring in modular architecture and enable developing/managing a large code base easier. The [blog post](https://logz.io/blog/opensearch-plugins/) from our partner [Logz.io](http://logz.io/) really helps understand why pluggable architecture is important and how the architecture works. 
+
+Plugin architecture is designed to enable solving specific problems and extending generic features. For example, [Anomaly Detection](https://github.com/opensearch-project/anomaly-detection) plugin reads time stream data ingested and finds anomalies. Another example is [Job Scheduler](https://github.com/opensearch-project/job-scheduler) plugin which schedules and runs generic jobs. 
+
+Plugins are of various types, generally could be categorized as:
+
+
+* Analysis: Used for analysis of data available within the cluster. 
+* Discovery: Used for easy discovery of nodes in various platforms.
+* Ingest: Used for pre/post processing data during ingestion.
+* Mappers: Helps extend/create data fields to OpenSearch.
+* Snapshot/Restore: Used to create snapshot and restore data
+
+To develop these plugins, the code base has well defined [interfaces](https://github.com/opensearch-project/OpenSearch/tree/main/server/src/main/java/org/opensearch/plugins) to solve specific sub-set of problems. 
+
+
+## Extension Points
+
+The architecture is built for plugins to hook onto various extension points with in the code base and subscribe to notifications/events they are interested in. There are a bunch of extension points few are default for all plugins and the rest are custom defined by few plugin interfaces. 
+
+Extension points enable plugins to hook into various events within the lifecycle of OpenSearch cluster. 
+The default extension points are defined by [Plugin.java](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/Plugin.java#L90) abstract class:
+
+
+* `getFeature` - Could be used to implement a custom feature and respond to Cluster state API.
+* `createGuiceModules` - Node level guice modules.
```

Comment:
Explain what guice is

Reply:
Its very hard to explain guice, but put it in generic terms.